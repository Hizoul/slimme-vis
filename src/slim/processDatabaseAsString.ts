
export interface IClass {
  label: string
  values: number
}

export interface Database {[index: string]: number[][] | undefined}

const processDatabaseAsString = (databaseContent: string) => {
  const classes: Database = {}
  for (const line of databaseContent.split("\n")) {
    const values = line.split(" ").map((value) => {
      if (value.length > 0) {
        return Number(value)
      }
    }).filter((val) => val != null)
    const classKey = String(values[0])
    if (classes[classKey] == null) {
      classes[classKey] = []
    }
    classes[classKey].push(values.splice(1))
  }
  return classes
}

export default processDatabaseAsString
