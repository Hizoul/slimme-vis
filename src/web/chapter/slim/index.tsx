import * as React from "react"
import { KeywordCodeTable, KeywordDatabase } from "../../components/keywords"

const SLIMChapter: React.FunctionComponent<any> = (props) => {
  return (
    <div className="c2 center">
      <h1>Slim</h1>
      <div className="line">
        In this chapter the SLIM algorithm will be explained step by step.
      </div>
      <div className="line">
        As input SLIM receives a <KeywordDatabase />
      </div>
      <div className="line">
        The output is a <KeywordCodeTable /> that tries to minimize the required amount of bits (MDL)
      </div>
    </div>
  )
}

export default SLIMChapter
