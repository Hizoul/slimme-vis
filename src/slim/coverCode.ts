import { Database } from "./processDatabaseAsString"
import { SortedCodeTable, subset } from "./standardCodeTable"

const coverCode = (codeTable: SortedCodeTable, setToUse: number[]) => {
  const set = setToUse.slice(0)
  const cover = []
  COVER: for (const code of codeTable) {
    if (subset(code[0], set).length === code[0].length) {
      cover.push(code[0])
      for (const value of code[0]) {
        set.splice(set.indexOf(value), 1)
      }
      if (set.length === 0) {
        break COVER
      }
    }
  }
  return cover
}

export default coverCode
