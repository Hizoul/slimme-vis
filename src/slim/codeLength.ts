import isEqual from "../util/isEqual"
import log2 from "../util/log"
import coverCode from "./coverCode"
import { Database } from "./processDatabaseAsString"
import { CodeTable, SortedCodeTable } from "./standardCodeTable"

const itemProbability = (codeTable: SortedCodeTable, set: number[]) => {
  let top = 0
  let fittingEntry: any
  let frequencySum = 0
  codeTable.forEach((entry) => {
    frequencySum += entry[1][0]
    if (isEqual(entry[0], set)) {
      fittingEntry = entry
    }
  })
  if (fittingEntry != null) {
    top = fittingEntry[1][0]
  }
  return top / frequencySum
}

const lengthForSet = (codeTable: SortedCodeTable, set: number[]) => {
  return -log2(itemProbability(codeTable, set))
}

const lengthForSetInStandard = (codeTable: SortedCodeTable, set: number[]) => {
  const cover = coverCode(codeTable, set)
  let coverSum = 0
  for (const code of cover) {
    coverSum += lengthForSet(codeTable, code)
  }
  return coverSum
}

const codeLength = (dataset: Database, codeTable: SortedCodeTable) => {
  let sum = 0
  for (const classKey of Object.keys(dataset)) {
    for (const itemSet of dataset[classKey]) {
      const cover = coverCode(codeTable, itemSet)
      let coverSum = 0
      for (const code of cover) {
        coverSum += lengthForSet(codeTable, code)
      }
      sum += coverSum
    }
  }
  return sum
}
function isNumber(n: number) {
  return typeof n === "number" && !isNaN(n) && isFinite(n)
}
const lengthWithStandard = (dataset: Database, codeTable: SortedCodeTable, standardTable: SortedCodeTable) => {
  let sum = 0
  for (const entry of codeTable) {
    const a = lengthForSet(codeTable, entry[0])
    const b = lengthForSet(standardTable, entry[0])
    if (isNumber(a)) {
      sum += a
    }
    if (isNumber(b)) {
      sum += b
    }
  }
  return sum
}

const codeLengthIncludingStandard = (dataset: Database, codeTable: SortedCodeTable, standardTable: SortedCodeTable) => {
  return lengthWithStandard(dataset, codeTable, standardTable) + codeLength(dataset, codeTable)
}

export default codeLength
export {
  codeLengthIncludingStandard, lengthForSet, lengthForSetInStandard
}
