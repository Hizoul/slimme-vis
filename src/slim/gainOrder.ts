import isEqual from "../util/isEqual"
import log2 from "../util/log"
import { lengthForSet, lengthForSetInStandard } from "./codeLength"
import { Database } from "./processDatabaseAsString"
import { ICodeTableEntry, SortedCodeTable, subset, updateFrequencies } from "./standardCodeTable"

export type ICandidatePair = number[][]

export type CandidatePairs = ICandidatePair[]

const createPairCandidates = (codeTable: SortedCodeTable) => {
  const pairs: ICandidatePair[] = []
  for (let i = 0; i < codeTable.length; i++) {
    for (let b = i; b < codeTable.length; b++) {
      const setA = codeTable[i][0]
      const setB = codeTable[b][0]
      if (i !== b && subset(setA, setB).length === 0) {
        const newCodeEntry: ICandidatePair = [setA, setB, [].concat(setA).concat(setB), [0]]
        pairs.unshift(newCodeEntry)
      }
    }
  }
  pairs.sort(sortPairsByLength)
  return pairs
}

const sortPairsByLength = (itemA: ICandidatePair, itemB: ICandidatePair) => {
  if (itemA[2].length < itemB[2].length) {
    return 1
  }
  if (itemA[2].length > itemB[2].length) {
    return -1
  }
  return -1 * JSON.stringify(itemA[2]).localeCompare(JSON.stringify(itemB[2]))
}

const getFrequencySum = (table: SortedCodeTable) => {
  let sum = 0
  for (const entry of table) {
    sum += entry[1][0]
  }
  return sum
}

const sortPairsByGain = (itemA: ICandidatePair, itemB: ICandidatePair) => {
  if (isNaN(itemA[3][0])) {
    return 1
  }
  if (isNaN(itemB[3][0])) {
    return -1
  }
  if (itemA[3][0] < itemB[3][0]) {
    return 1
  }
  if (itemA[3][0] > itemB[3][0]) {
    return -1
  }
  return -1 * JSON.stringify(itemA[3][0]).localeCompare(JSON.stringify(itemB[3][0]))
}

const shuffleArray = (array: any[]) => {
  for (let i = array.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1))
    const temp = array[i]
    array[i] = array[j]
    array[j] = temp
  }
}

const gainOrder = (dataset: Database, currentTable: SortedCodeTable, standardCodeTable: SortedCodeTable, candidatePairs: CandidatePairs, sortOrder: number = 0) => {
  const s = getFrequencySum(currentTable)
  let i = 0
  let start = new Date().getTime()
  const xTidHolder: any = []
  const yTidHolder: any = []
  const tidHolder: any = []
  const clonedDb: Database = JSON.parse(JSON.stringify(dataset))
  for (const entry of currentTable) {
    entry[1][0] = 0
    for (const classKey of Object.keys(clonedDb)) {
      for (const transactionIndex in clonedDb[classKey]) {
        const transaction = clonedDb[classKey][transactionIndex]
        if (subset(entry[0], transaction).length === entry[0].length) {
          entry[1][0]++
          const key = JSON.stringify(entry[0])
          if (tidHolder[key] == null) {
            tidHolder[key] = []
          }
          tidHolder[key].push(Number(transactionIndex))
          clonedDb[classKey][transactionIndex] = transaction.filter((value) => entry[0].indexOf(value) === -1)
        }
      }
    }
  }
  const possiblePairs = []
  for (const pair of candidatePairs) {
    start = new Date().getTime()
    let xTids = tidHolder[JSON.stringify(pair[0])]
    if (xTids == null) {
      xTids = []
    }
    let yTids = tidHolder[JSON.stringify(pair[1])]
    if (yTids == null) {
      yTids = []
    }
    const xyn = subset(xTids, yTids).length
    if (xyn === 0) {

    } else {
      const xn = xTids.length - xyn
      const yn = yTids.length - xyn
      const sn = s - xyn
      // gain = L(D|CT) + L(CT, D)
      // L(D|CT) = encoding bit difference
      // clogcres == sum(c * log c- c'* log c')
      let clogcres = 0
      // cdiff == sum log c'-log c
      let cdiff = 0
      // sum log c' - l(C|ST)
      let cnst = 0
      // sum log c - l(C|ST)
      let cst = 0
      for (let i = 0; i < currentTable.length; i++) {
        const C = currentTable[i][0]
        const c = currentTable[i][1][0]
        let cn = currentTable[i][1][0]
        if (isEqual(currentTable[i][0], pair[0])) {
          cn = xn
        } else if (isEqual(currentTable[i][0], pair[1])) {
          cn = yn
        }
        if (cn !== c) {
          if (c !== 0 && cn !== 0) {
            clogcres += c * log2(c) - cn * log2(cn)
            cdiff += log2(cn) - log2(c)
          }
          const lengthInStandard = lengthForSetInStandard(standardCodeTable, C)
          if (c === 0) {
            cnst += log2(cn) - lengthInStandard
          }
          if (cn === 0) {
            cst += lengthInStandard - log2(c)
          }
        }
      }
      const bitDifference = (s * log2(s) - sn * log2(sn)) + (xyn * log2(xyn)) - clogcres
      const complexityDifference = log2(xyn) - lengthForSetInStandard(standardCodeTable, pair[2])
      + currentTable.length * log2(s) - (currentTable.length + 1) * log2(sn) + cdiff + cnst + cst
      pair[3][0] = bitDifference + complexityDifference
      possiblePairs.push(pair)
    }
    i++
  }
  console.log("GOT SORT ORDER", sortOrder)
  switch (sortOrder) {
    case 0: {
      possiblePairs.sort(sortPairsByGain).reverse()
      break
    }
    case 1: {
      possiblePairs.sort(sortPairsByGain)
      break
    }
    case 2: {
      shuffleArray(possiblePairs)
      break
    }
  }
  return possiblePairs
}

export default gainOrder
export {
  createPairCandidates
}
