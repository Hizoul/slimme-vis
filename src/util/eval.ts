import { codeLengthIncludingStandard } from "../slim/codeLength"
import { Database } from "../slim/processDatabaseAsString"
import sortedStandardCodeTable, { SortedCodeTable } from "../slim/standardCodeTable"

const getLPercent = (dataset: Database, codeTable: SortedCodeTable) => {
  const ST = sortedStandardCodeTable(dataset)
  return (codeLengthIncludingStandard(dataset, codeTable, ST) / codeLengthIncludingStandard(dataset, ST, ST)) * 100
}
// L%= L(D,CT)Ã—100, L(D, ST )
export {
  getLPercent
}
