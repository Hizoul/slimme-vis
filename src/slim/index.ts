import { getLPercent } from "../util/eval"
import isValidTable from "../util/isValidTable"
import { codeLengthIncludingStandard } from "./codeLength"
import gainOrder, { createPairCandidates } from "./gainOrder"
import postAcceptancePruning from "./postAcceptancePruning"
import { Database } from "./processDatabaseAsString"
import standardCodeTable, { SortedCodeTable, supportOfItem, updateFrequencies } from "./standardCodeTable"

// The function should do the following
// CT ← Standard Code Table(D)
// forF∈{X∪Y:X,Y∈CT}in GainOrder do
//   CTc ← (CT ⊕ F ) in Standard Cover Order
//   if L(D,CTc) < L(D,CT) then
//     CT ← post-prune(CTc) end if
// end for return
const slim = (dataset: Database, tableToExplore?: SortedCodeTable, sortType?: number) => {
  // CT ← Standard Code Table(D)
  const ST = standardCodeTable(dataset)
  let improved = true
  let CT = tableToExplore ? tableToExplore : ST
  let i = 0
  const start = Date.now()
  while (improved) {
    console.log(`in iteration ${i} of slim current |L%|=${getLPercent(dataset, CT)} elapsed time ${Date.now() - start}ms`)
    i++
    improved = false
    const candidates = createPairCandidates(CT)
    const sortedCandidates = gainOrder(dataset, CT, ST, candidates, sortType)

    // for F∈{X∪Y:X,Y∈CT} in GainOrder do
    PROV: for (const candidate of sortedCandidates) {
      // CTc ← (CT ⊕ F ) in Standard Cover Order
      let CTc = JSON.parse(JSON.stringify(CT))
      CTc.unshift([candidate[2], [0], [supportOfItem(candidate[2], dataset)]])
      CTc = updateFrequencies(dataset, CTc)
      // if L(D,CTc) < L(D,CT) then
      if (codeLengthIncludingStandard(dataset, CTc, ST) < codeLengthIncludingStandard(dataset, CT, ST)) {
        // CT ← post-prune(CTc)
        CTc = postAcceptancePruning(dataset, CT, CTc, ST)
        CT = CTc
        improved = true
        break PROV
      }
    }
  }
  console.log("DONE WITH SLIM")
  return CT
}

export default slim
