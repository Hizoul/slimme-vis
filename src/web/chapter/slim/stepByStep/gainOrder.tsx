import * as React from "react"
import { FaArrowRight } from "react-icons/fa"
import gainOrder, { createPairCandidates, ICandidatePair } from "../../../../slim/gainOrder"
import { exampleDataset, exampleStandardCodeTable, exampleTranslation } from "../../../../util/exampleData"
import translate from "../../../../util/translation"
import CodeTable from "../../../components/codeTable"
import { KeywordCodeTable } from "../../../components/keywords"
import { bigIconClasses } from "../../globals"

const PairDisplay: React.FunctionComponent<{
  candidates: ICandidatePair[], showGain?: boolean
}> = (props) => {
  return (
    <div>
      {props.candidates.map((set) =>
        <div key={JSON.stringify(set[2])}>
        {props.showGain ? <span>Gain: {set[3][0].toFixed(2)};&nbsp;</span> : null}
          {translate(set[2], exampleTranslation)}
        </div>
      )}
    </div>
  )
}

const SLIMGainOrder: React.FunctionComponent<any> = (props) => {
  const orderedByGain = gainOrder(exampleDataset, exampleStandardCodeTable,
    exampleStandardCodeTable, createPairCandidates(exampleStandardCodeTable))
  return (
    <div className="c2">
      <div className="line">
        The second step, is, where SLIM improves upon KRIMP. Instead of a predefined list of candidates <i>F</i>, it builds them on the fly, based on the current <KeywordCodeTable />.
      </div>
      <div className="inline">
        <CodeTable codeTable={exampleStandardCodeTable} translation={exampleTranslation} />
        <div className="center">
          <span>Build all pairwise combinations from the sets of <i>CT</i></span>
          <br />
          <FaArrowRight className={bigIconClasses} />
        </div>
        <div style={{minWidth: "15%", textAlign: "center"}}>
          <div className="limitHeight">
            <PairDisplay candidates={createPairCandidates(exampleStandardCodeTable)} />
          </div>
        </div>
          <div className="center">
            Calculate estimated gain and sort by it
            <br />
            <FaArrowRight className={bigIconClasses} />
          </div>
        <div>
          <div className="limitHeight">
            <PairDisplay candidates={orderedByGain} showGain={true} />
          </div>
        </div>
      </div>
    </div>
  )
}

export default SLIMGainOrder
