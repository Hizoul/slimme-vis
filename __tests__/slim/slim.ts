import slim from "../../src/slim"
import gainOrder, { createPairCandidates } from "../../src/slim/gainOrder"
import processDatabaseAsString from "../../src/slim/processDatabaseAsString"
import sortedStandardCodeTable, { SortedCodeTable, updateFrequencies } from "../../src/slim/standardCodeTable"
import { getLPercent } from "../../src/util/eval"

const database = `
0 1 2 3
0 1 2 3
0 1 2 3
0 1 2 3
0 1 2 3
0 1 2
0 1
0 2
`
describe("GainOrder", () => {
  it("should sort correctly", () => {
    const dataset = processDatabaseAsString(database)
    const slimTable = slim(dataset)
    expect(slimTable).toMatchSnapshot("optimized table")
    expect(getLPercent(dataset, slimTable)).toMatchSnapshot("encoindg size new")
    expect(getLPercent(dataset, slimTable) < getLPercent(dataset, sortedStandardCodeTable(dataset))).toBeTruthy()
  })
})
