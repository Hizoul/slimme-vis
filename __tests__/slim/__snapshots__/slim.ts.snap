// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GainOrder should sort correctly: encoindg size new 1`] = `53.25412437118968`;

exports[`GainOrder should sort correctly: optimized table 1`] = `
Array [
  Array [
    Array [
      2,
      3,
      1,
    ],
    Array [
      5,
    ],
  ],
  Array [
    Array [
      2,
    ],
    Array [
      2,
    ],
  ],
  Array [
    Array [
      1,
    ],
    Array [
      2,
    ],
  ],
]
`;
